File: games/lynrummy/python/bench_outer_shell.py
#!/usr/bin/env python3
"""bench_outer_shell.py — Compare outer-shell modes on random hands.

Fixed corpus: 60 random 6-card hands drawn from the 81 cards not on
the Game 17 opening board (6 helpers, 23 cards), seed 42.

NOTE: 6-card hands are used here solely to exercise all four outcome
types (triple, pair, single, stuck) within a manageable corpus. Real
Lyn Rummy hands start at 15 cards; hand size shrinks as cards are
played to the board.

Two modes compared:

  singleton-only  skip pair/triple steps; project each hand card as a
                  singleton trouble, pick the shortest BFS plan.

  full            triple-in-hand first (no BFS), then every valid pair
                  as a 2-partial trouble, then every singleton; pick
                  shortest plan overall. This is agent_prelude.find_play.

Usage:
  python3 bench_outer_shell.py
"""

import random
import time
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

import bfs
import agent_prelude
from rules import classify, is_partial_ok
from rules.card import card as make_card, card_label

_N = 60
_HAND_SIZE = 6   # benchmark only — not actual gameplay hand size
_SEED = 42
_MAX_STATES = 5000

# ── Fixed board ──────────────────────────────────────────────────────────────

_BOARD_STACKS_LABELS = [
    ["KS", "AS", "2S", "3S"],
    ["TD", "JD", "QD", "KD"],
    ["2H", "3H", "4H"],
    ["7S", "7D", "7C"],
    ["AC", "AD", "AH"],
    ["2C", "3D", "4C", "5H", "6S", "7H"],
]


def _make_board():
    return [[make_card(lbl, deck=0) for lbl in stack]
            for stack in _BOARD_STACKS_LABELS]


def _remaining_cards():
    """The 81 cards not on the board."""
    on_board = {make_card(lbl, deck=0)
                for stack in _BOARD_STACKS_LABELS for lbl in stack}
    out = []
    for si in range(4):
        for vi in range(13):
            for deck in (0, 1):
                c = (vi + 1, si, deck)
                if c not in on_board:
                    out.append(c)
    assert len(out) == 81
    return out


# ── Singleton-only mode ───────────────────────────────────────────────────────

def _project_singleton(board, c):
    """BFS with `c` as a singleton trouble. Returns (plan_or_None, ms)."""
    augmented = board + [[c]]
    helper = [s for s in augmented if classify(s) != "other"]
    trouble = [s for s in augmented if classify(s) == "other"]
    t0 = time.perf_counter()
    plan = bfs.solve_state_with_descs(
        (helper, trouble, [], []),
        max_trouble_outer=10,
        max_states=_MAX_STATES,
        verbose=False,
    )
    return plan, (time.perf_counter() - t0) * 1000


def find_play_singletons_only(hand, board):
    """Singleton-only: try every hand card as a trouble singleton,
    return the one with the shortest BFS plan (or None if stuck)."""
    candidates = []
    total_ms = 0.0
    projections = 0
    for c in hand:
        plan, ms = _project_singleton(board, c)
        total_ms += ms
        projections += 1
        if plan is not None:
            candidates.append({"placements": [c], "plan": plan})
    result = (None if not candidates
              else min(candidates, key=lambda r: len(r["plan"])))
    return result, total_ms, projections


# ── Full mode ─────────────────────────────────────────────────────────────────

def find_play_full(hand, board):
    """Full mode via agent_prelude.find_play_with_budget."""
    t0 = time.perf_counter()
    result = agent_prelude.find_play_with_budget(
        hand, board, max_states=_MAX_STATES
    )
    total_ms = (time.perf_counter() - t0) * 1000
    # Rough projection count: valid pairs + singletons; for display only.
    n_pairs = sum(
        1 for i, c1 in enumerate(hand)
        for c2 in hand[i + 1:]
        if is_partial_ok([c1, c2])
    )
    projections = n_pairs + len(hand)
    return result, total_ms, projections


# ── Formatting helpers ────────────────────────────────────────────────────────

def _fmt_result(result):
    if result is None:
        return "stuck"
    placements = " ".join(card_label(c) for c in result["placements"])
    n = len(result["plan"])
    kind = "pair" if len(result["placements"]) == 2 else (
        "triple" if len(result["placements"]) == 3 else "single")
    return f"{kind} [{placements}] → {n}-step plan"


def _plan_len(result):
    return 999 if result is None else len(result["plan"])


def _placement_count(result):
    return 0 if result is None else len(result["placements"])


def _outcome(result):
    if result is None:
        return "stuck"
    n = len(result["placements"])
    if n >= 3:
        return "triple"
    if n == 2:
        return "pair"
    return "single"


# ── Main ──────────────────────────────────────────────────────────────────────

def main():
    remaining = _remaining_cards()
    rng = random.Random(_SEED)
    hands = [rng.sample(remaining, _HAND_SIZE) for _ in range(_N)]
    board = _make_board()

    print(f"Game 17 board  ·  {_N} hands of {_HAND_SIZE} (benchmark size)  ·  "
          f"seed={_SEED}  ·  max_states={_MAX_STATES}")
    print()

    col = 44  # width for result column

    # ── singleton-only pass ──────────────────────────────────────────────
    print("=== singleton-only (no pair/triple) ===")
    solo_times = []
    solo_results = []
    for i, hand in enumerate(hands):
        result, ms, n_proj = find_play_singletons_only(hand, board)
        solo_times.append(ms)
        solo_results.append(result)
        desc = _fmt_result(result)
        print(f"  hand {i+1:2}  {desc:<{col}}  {ms:7.1f}ms  ({n_proj} projections)")
    solo_total = sum(solo_times)
    solo_stuck = sum(1 for r in solo_results if r is None)
    print(f"  ── total {solo_total:.0f}ms  ·  stuck {solo_stuck}/{_N}\n")

    # ── full pass ────────────────────────────────────────────────────────
    print("=== full (triple-in-hand + pair-BFS + singleton) ===")
    full_times = []
    full_results = []
    for i, hand in enumerate(hands):
        result, ms, n_proj = find_play_full(hand, board)
        full_times.append(ms)
        full_results.append(result)
        desc = _fmt_result(result)
        print(f"  hand {i+1:2}  {desc:<{col}}  {ms:7.1f}ms  (~{n_proj} projections)")
    full_total = sum(full_times)
    full_stuck = sum(1 for r in full_results if r is None)
    print(f"  ── total {full_total:.0f}ms  ·  stuck {full_stuck}/{_N}\n")

    # ── per-hand comparison ───────────────────────────────────────────────
    better_plan = 0
    same_plan = 0
    worse_plan = 0
    more_placements = 0
    for rs, rf in zip(solo_results, full_results):
        sp, fp = _plan_len(rs), _plan_len(rf)
        sc, fc = _placement_count(rs), _placement_count(rf)
        if fp < sp:
            better_plan += 1
        elif fp == sp:
            same_plan += 1
        else:
            worse_plan += 1
        if fc > sc:
            more_placements += 1

    # ── outcome coverage (full mode) ────────────────────────────────────
    outcomes = [_outcome(r) for r in full_results]
    counts = {k: outcomes.count(k) for k in ("triple", "pair", "single", "stuck")}

    # ── summary ──────────────────────────────────────────────────────────
    ratio = full_total / max(solo_total, 0.001)
    print("=== summary ===")
    print(f"  singleton-only  {solo_total:7.0f}ms total  stuck {solo_stuck}/{_N}")
    print(f"  full            {full_total:7.0f}ms total  stuck {full_stuck}/{_N}")
    print(f"  wall ratio (full/solo): {ratio:.2f}x", end="")
    if ratio > 1:
        print(f"  (full is {(ratio-1)*100:.0f}% slower in wall time)")
    else:
        print(f"  (full is {(1-ratio)*100:.0f}% faster in wall time)")
    print(f"  plan improvement: better={better_plan}  same={same_plan}"
          f"  worse={worse_plan}  more-placements={more_placements}"
          f"  (out of {_N} hands)")
    print(f"  outcome coverage (full): "
          f"triple={counts['triple']}  pair={counts['pair']}  "
          f"single={counts['single']}  stuck={counts['stuck']}")


if __name__ == "__main__":
    main()

File: games/lynrummy/ts/bench/bench_outer_shell.ts
// bench_outer_shell.ts — Compare outer-shell modes on random hands.
//
// Fixed corpus: 60 random 6-card hands drawn from the 81 cards not on
// the Game 17 opening board (6 helpers, 23 cards), seed 42 — see
// `baseline_deal.ts` for the canonical PRNG + deal.
//
// Two modes compared:
//
//   singleton-only  skip pair/triple steps; project each hand card as a
//                   singleton trouble, pick the shortest BFS plan.
//
//   full            triple-in-hand first (no BFS), then every valid pair
//                   as a 2-partial trouble, then every singleton; pick
//                   shortest plan overall. This is hand_play.findPlay.
//
// Usage:
//   node bench/bench_outer_shell.ts

import { type Card, cardLabel } from "../core/card.ts";
import { isPartialOk } from "../core/card_stack.ts";
import { solveBoard } from "../bfs/engine_v2.ts";
import { findPlay, type LogicalMovesForPlay } from "../step/hand_play.ts";
import {
  openingBoardCardLists,
  remainingCards,
  mulberry32,
  shuffle,
} from "../baseline_deal.ts";

const N_HANDS = 60;
const HAND_SIZE = 6;
const SEED = 42;

// Per-hand min-of-N timing parameters. Single-shot is too noisy
// (individual swings 30-200% on a loaded system); min-of-N with a
// warmup stabilizes the gold so it can serve as a real timing
// trip-wire, not just a snapshot.
//
// MIN_OF_N = 5 picked to give stable totals at acceptable wall time
// (~15-20s total for the full bench). Increase if the gold's times
// still bounce between runs.
const TIMING_WARMUP_RUNS = 1;
const TIMING_MIN_OF_N = 5;

function timeMinOfN<T>(work: () => T): { result: T; bestMs: number } {
  for (let i = 0; i < TIMING_WARMUP_RUNS; i++) work();
  let bestMs = Infinity;
  let result!: T;
  for (let i = 0; i < TIMING_MIN_OF_N; i++) {
    const t0 = performance.now();
    result = work();
    const ms = performance.now() - t0;
    if (ms < bestMs) bestMs = ms;
  }
  return { result, bestMs };
}

// ── Singleton-only mode ─────────────────────────────────────────────
//
// These return just { result, projections }; timing is the caller's
// responsibility (via timeMinOfN). Keeping the work pure of timing
// makes it easy to wrap in min-of-N without per-call instrumentation.

function projectSingleton(
  board: readonly (readonly Card[])[],
  c: Card,
): LogicalMovesForPlay | null {
  const augmented = [...board, [c]];
  const result = solveBoard(augmented);
  if (result === null) return null;
  const moves = result.plan.map(p => p.move);
  const moveLines = result.plan.map(p => p.line);
  return { cardsToPlay: [c], moves, moveLines };
}

interface SingletonResult {
  result: LogicalMovesForPlay | null;
  projections: number;
}

function findPlaySingletonsOnly(
  hand: readonly Card[],
  board: readonly (readonly Card[])[],
): SingletonResult {
  const candidates: LogicalMovesForPlay[] = [];
  for (const c of hand) {
    const r = projectSingleton(board, c);
    if (r !== null) candidates.push(r);
  }
  if (candidates.length === 0) return { result: null, projections: hand.length };
  const result = candidates.reduce((best, cur) =>
    cur.moves.length < best.moves.length ? cur : best,
  );
  return { result, projections: hand.length };
}

// ── Full mode ───────────────────────────────────────────────────────

interface FullResult {
  result: LogicalMovesForPlay | null;
  projections: number;
}

function findPlayFull(
  hand: readonly Card[],
  board: readonly (readonly Card[])[],
): FullResult {
  const result = findPlay(hand, board);
  // Rough projection count for display: valid pairs + singletons.
  let nPairs = 0;
  for (let i = 0; i < hand.length; i++) {
    for (let j = i + 1; j < hand.length; j++) {
      if (isPartialOk([hand[i]!, hand[j]!])) nPairs++;
    }
  }
  return { result, projections: nPairs + hand.length };
}

// ── Formatting ──────────────────────────────────────────────────────

function fmtResult(result: LogicalMovesForPlay | null): string {
  if (result === null) return "stuck";
  const labels = result.cardsToPlay.map(cardLabel).join(" ");
  const n = result.moves.length;
  const kind =
    result.cardsToPlay.length === 2
      ? "pair"
      : result.cardsToPlay.length === 3
        ? "triple"
        : "single";
  return `${kind} [${labels}] → ${n}-step plan`;
}

function planLen(r: LogicalMovesForPlay | null): number {
  return r === null ? 999 : r.moves.length;
}

function placementCount(r: LogicalMovesForPlay | null): number {
  return r === null ? 0 : r.cardsToPlay.length;
}

function outcome(r: LogicalMovesForPlay | null): "stuck" | "triple" | "pair" | "single" {
  if (r === null) return "stuck";
  const n = r.cardsToPlay.length;
  if (n >= 3) return "triple";
  if (n === 2) return "pair";
  return "single";
}

function pad(s: string, width: number): string {
  return s.length >= width ? s : s + " ".repeat(width - s.length);
}

function fmtMs(ms: number): string {
  return ms.toFixed(1).padStart(7, " ");
}

// ── Main ────────────────────────────────────────────────────────────

function main(): void {
  const remaining = remainingCards();
  const rng = mulberry32(SEED);
  const hands: Card[][] = [];
  for (let i = 0; i < N_HANDS; i++) hands.push(shuffle(remaining, rng).slice(0, HAND_SIZE));
  const board = openingBoardCardLists();

  console.log(
    `Game 17 board  ·  ${N_HANDS} hands of ${HAND_SIZE} (benchmark size)  ·  seed=${SEED}`,
  );
  console.log();

  const col = 44;

  // Singleton-only pass — min-of-N per hand.
  console.log("=== singleton-only (no pair/triple) ===");
  const soloTimes: number[] = [];
  const soloResults: (LogicalMovesForPlay | null)[] = [];
  for (let i = 0; i < hands.length; i++) {
    const { result: solo, bestMs } = timeMinOfN(() => findPlaySingletonsOnly(hands[i]!, board));
    soloTimes.push(bestMs);
    soloResults.push(solo.result);
    const desc = pad(fmtResult(solo.result), col);
    console.log(`  hand ${String(i + 1).padStart(2, " ")}  ${desc}  ${fmtMs(bestMs)}ms  (${solo.projections} projections)`);
  }
  const soloTotal = soloTimes.reduce((a, b) => a + b, 0);
  const soloStuck = soloResults.filter(r => r === null).length;
  console.log(`  ── total ${soloTotal.toFixed(0)}ms  ·  stuck ${soloStuck}/${N_HANDS}\n`);

  // Full pass — min-of-N per hand.
  console.log("=== full (triple-in-hand + pair-BFS + singleton) ===");
  const fullTimes: number[] = [];
  const fullResults: (LogicalMovesForPlay | null)[] = [];
  for (let i = 0; i < hands.length; i++) {
    const { result: full, bestMs } = timeMinOfN(() => findPlayFull(hands[i]!, board));
    fullTimes.push(bestMs);
    fullResults.push(full.result);
    const desc = pad(fmtResult(full.result), col);
    console.log(`  hand ${String(i + 1).padStart(2, " ")}  ${desc}  ${fmtMs(bestMs)}ms  (~${full.projections} projections)`);
  }
  const fullTotal = fullTimes.reduce((a, b) => a + b, 0);
  const fullStuck = fullResults.filter(r => r === null).length;
  console.log(`  ── total ${fullTotal.toFixed(0)}ms  ·  stuck ${fullStuck}/${N_HANDS}\n`);

  // Per-hand comparison.
  let betterPlan = 0,
    samePlan = 0,
    worsePlan = 0,
    morePlacements = 0;
  for (let i = 0; i < N_HANDS; i++) {
    const sp = planLen(soloResults[i]!);
    const fp = planLen(fullResults[i]!);
    const sc = placementCount(soloResults[i]!);
    const fc = placementCount(fullResults[i]!);
    if (fp < sp) betterPlan++;
    else if (fp === sp) samePlan++;
    else worsePlan++;
    if (fc > sc) morePlacements++;
  }

  const counts = { triple: 0, pair: 0, single: 0, stuck: 0 };
  for (const r of fullResults) counts[outcome(r)]++;

  const ratio = fullTotal / Math.max(soloTotal, 0.001);
  console.log("=== summary ===");
  console.log(`  singleton-only  ${soloTotal.toFixed(0).padStart(7, " ")}ms total  stuck ${soloStuck}/${N_HANDS}`);
  console.log(`  full            ${fullTotal.toFixed(0).padStart(7, " ")}ms total  stuck ${fullStuck}/${N_HANDS}`);
  const ratioLine = `  wall ratio (full/solo): ${ratio.toFixed(2)}x`;
  if (ratio > 1) {
    console.log(`${ratioLine}  (full is ${((ratio - 1) * 100).toFixed(0)}% slower in wall time)`);
  } else {
    console.log(`${ratioLine}  (full is ${((1 - ratio) * 100).toFixed(0)}% faster in wall time)`);
  }
  console.log(
    `  plan improvement: better=${betterPlan}  same=${samePlan}  worse=${worsePlan}  more-placements=${morePlacements}  (out of ${N_HANDS} hands)`,
  );
  console.log(
    `  outcome coverage (full): triple=${counts.triple}  pair=${counts.pair}  single=${counts.single}  stuck=${counts.stuck}`,
  );
}

main();
